packages/js/src/plugins/typescript/plugin.ts

```diff
@@ -622,30 +622,39 @@ function isValidPackageJsonBuildConfig(
     joinPathFragments(projectRoot, 'package.json')
   );
 
-  const rootDir = tsConfig.options.rootDir ?? 'src/';
-  if (!tsConfig.options.rootDir) {
-    console.warn(
-      `The 'rootDir' option is not set in the tsconfig file at ${tsConfigPath}. Assuming 'src/' as the root directory.`
-    );
-  }
+  const rootDir = tsConfig.options.rootDir ?? 'src';
+  const resolvedRootDir = resolve(workspaceRoot, projectRoot, rootDir);
 
-  const isPathWithinSrc = (path: string): boolean => {
-    const resolvedRootDir = resolve(workspaceRoot, projectRoot, rootDir);
+  const outDir = tsConfig.options.outDir
+    ? tsConfig.options.outDir
+    : tsConfig.options.outFile
+    ? dirname(tsConfig.options.outFile)
+    : 'src'; // tsconfig defaults to 'src' if outDir and outFile are not set
+
+  const isPathSourceFile = (path: string): boolean => {
+    const ext = extname(path);
+
+    const resolvedOutDir = resolve(workspaceRoot, projectRoot, outDir);
     const pathToCheck = resolve(workspaceRoot, projectRoot, path);
 
-    return pathToCheck.startsWith(resolvedRootDir);
+    if (resolvedRootDir === resolvedOutDir) {
+      // Check that the file extension is `.ts`, `.cts`, or `.mts`. As the source files are in the same directory as the output files.
+      return ['.ts', '.tsx', '.cts', '.mts'].includes(ext);
+    } else {
+      return !pathToCheck.startsWith(resolvedOutDir);
+    }
   };
 
   // If `outFile` is defined, check the validity of the path.
   if (tsConfig.options.outFile) {
-    if (isPathWithinSrc(tsConfig.options.outFile)) {
+    if (isPathSourceFile(tsConfig.options.outFile)) {
       return false;
     }
   }
 
   const buildPaths = ['main', 'module'];
```

Comment:
We should check `exports` before `main` and `module`. From Node docs:

> When a package has an ["exports"](https://nodejs.org/docs/latest-v22.x/api/packages.html#exports) field, this will take precedence over the "main" field when importing the package by name.

See: https://nodejs.org/docs/latest-v22.x/api/packages.html#main